hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java

```diff
@@ -0,0 +1,498 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.operation.steps.engine;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.core.EntryEventType;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.util.Clock;
+import com.hazelcast.map.EntryProcessor;
+import com.hazelcast.map.impl.MapEntries;
+import com.hazelcast.map.impl.operation.EntryOperator;
+import com.hazelcast.map.impl.operation.MapOperation;
+import com.hazelcast.map.impl.record.Record;
+import com.hazelcast.map.impl.recordstore.RecordStore;
+import com.hazelcast.map.impl.recordstore.StaticParams;
+import com.hazelcast.query.Predicate;
+import com.hazelcast.query.impl.InternalIndex;
+import com.hazelcast.spi.merge.SplitBrainMergePolicy;
+import com.hazelcast.spi.merge.SplitBrainMergeTypes;
+import com.hazelcast.wan.impl.CallerProvenance;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.UUID;
+
+import static com.hazelcast.map.impl.record.Record.UNSET;
+
+/**
+ * {@link Step} uses this state object to
+ * transfer state between subsequent steps.
+ */
+@SuppressWarnings("checkstyle:methodcount")
+public class State {
```

Comment:
The fact that this class has too many fields gives me the feeling that it will create a litter. But especially the benchmark result, which Frantisek requested, shows that currently this doesn't have a significant effect. I'm writing to note if the bottleneck slides here later -> keeping these state fields in a Map<String, Object> can reduce the memory usage stems from by these fields, most of which are optional.

Reply:
We will lose type checks in that case if we make use of `Map<String, Object>`.
Now in this simple form it works as intended, i will keep it as is for now.